OctopusDeployLabs/terraform-provider-octopusdeploy
# Isaac/plugin framework testing

This PR includes a number of updates to our testing setup and introduces the SDK and Framework testing approaches to our CI coverage

[Results](https://github.com/OctopusDeployLabs/terraform-provider-octopusdeploy/runs/27297439008) - Note this run was significantly slower than the usual ~10 minutes I've been seeing.

Over the years the terraform provider has had a few different testing approaches
* Terratest
* SDK testing
* Integration testing - using https://github.com/OctopusSolutionsEngineering/OctopusTerraformTestFramework

At the time of this change only the integration tests were running in CI, SDK testing was not but mostly passed and Terratest was outdated to a point that most of the configuration examples either didn't exist or were failing. 

This PR does a few things 
* Run all tests across the solution
* Run on a shared container in CI, instead of creating a container per test. (Most tests get their own space)
* Update or skip any SDK tests that were previously failing
* Introduce Framework test example for the new provider 

Ideal changes for the future
* Update integration tests so they can also target a local instace
* Add support for setting the octoClient in the testing_containers.go setups to support local assertions.
* Use the migration tests, an example is under migration > testing > exampletest. This covers the scenario of testing a configuration that has run on the old provider being run against the new provider. The terraformrc override makes this a bit tricky.
* **Update containers to pull latest main builds (maybe a day or two delayed) from artifactory instead of latest LTS from Dockerhub. This should help coverage of newer breaking changes to resources + the API. (This is the most important imo)**

## Important notes

If you are looking to run tests locally, these will default to running against your local instance. This is only possible for the non integration tests, to run tests against containers (supported for all tests) update the _testing_containers_test.go_ `createSharedContainer` flag to true. This is required for each of the octopusdeploy and octopusdeploy_framework packages. There's a bit of future work around this that I would like to update.

**Please check the framework_provider.go changes.**
